# Feat: 뉴스 API 연동

xoals25 opened the pull request:
## 작업 내용 (Content)
- 네이버 검색 API를 통해 뉴스 데이터를 가져와서 상세 데이터를 크롤링 합니다. 크롤링한 데이터를 파싱하여 Bud 뉴스 데이터 형식에 맞게 저장합니다.

- 뉴스 목록과, 뉴스 상세 데이터를 조회할 수 있습니다.

## Merge 전 필요 작업 (Checklist before merge)

## 희망 리뷰 완료 일 (Expected due date)

xoals25 committed:
# Feat: 뉴스 API 연동

- 네이버 검색 API를 통해 뉴스 데이터를 가져와서 상세 데이터를 크롤링 합니다.
크롤링한 데이터를 파싱하여 Bud 뉴스 데이터 형식에 맞게 저장합니다.

- 뉴스 목록과, 뉴스 상세 데이터를 조회할 수 있습니다.

Ggyumalang commented on user-api/src/main/java/zerobase/bud/news/controller/NewsController.java:
```diff
@@ -0,0 +1,51 @@
+                DetailNewsResponse.from(newsService.getNewsDetail(id)));
+    }
+
+    @ExceptionHandler(DateTimeParseException.class)
+    public ErrorResponse handleDateTimeParseException(
+            DateTimeParseException e) {
+        log.info("INVALID_DATE_FORMAT (400 BAD_REQUEST) :: {}", e.getMessage());
+
+        return new ErrorResponse(ErrorCode.DATA_TYPE_UN_MATCH, ErrorCode.DATA_TYPE_UN_MATCH.getDescription());
```
오호 요로케 ExceptionHandling 하면 저번에 저희 회의 때 말씀하셨던 그 문제가 해결이 되는 것인가요?
요거 ResponseEntity<ErrorResponse> 로 감싸고 status를 ResponseEntity.badrequest().body(ErrorResponse)하면 status도 보여줄 수 있어서 괜찮을 것 같아 제안드립니다!

xoals25 replied:
아 이부분은 해결이 되지 않았습니다!
제가 지우고 올린다는걸 깜빡했습니다!

Ggyumalang commented on user-api/src/main/java/zerobase/bud/news/domain/News.java:
```diff
@@ -0,0 +1,32 @@
+    private String title;
+    private String link;
+    private String summaryContent;
+    @Lob
+    private String content;
```
대용량 Text가 들어올 것을 생각하셔서 Lob 넣으신 거 너무 좋은 것 같습니다!

thdefn replied:
https://ttl-blog.tistory.com/117 오 이런게 있네요 !!! 알아갑니다 !

xoals25 replied:
오오 제가 도움이되서 좋습니다!!
쓰는게 맞는건가 고민했었는데 다행이에요!

Ggyumalang commented on user-api/src/main/java/zerobase/bud/news/dto/SearchAllNews.java:
```diff
@@ -0,0 +1,86 @@
+    @Setter
+    @AllArgsConstructor
+    public static class Request {
+        String keyword;
+        @Nullable
+        int size;
+        @Nullable
+        int page;
+        NewsSortType sort;
+        @DateTimeFormat(iso = DateTimeFormat.ISO.DATE)
+        LocalDate startDate;
+        @DateTimeFormat(iso = DateTimeFormat.ISO.DATE)
+        LocalDate endDate;
```
요기 private가 빠진 것 같습니다. 혹시.. 이유가 있으신가요?

xoals25 replied:
제가 까먹고 빼먹었습니다!
꼼꼼하게 봐주셔서 너무 감사합니다! 수정 하겠습니다!

Ggyumalang commented on user-api/src/main/java/zerobase/bud/news/service/HttpService.java:
```diff
@@ -0,0 +1,62 @@
+
+            return responseBody.toString();
+        } catch (IOException e) {
+            throw new RuntimeException("API 응답을 읽는데 실패했습니다.", e);
```
요 부분 혹시 BudException으로 하면 BudException으로 Exception이 안떨어져서 RuntimeException 주신 건가요?

xoals25 replied:
이 부분은 HttpService 연결할 때 사용하는 코드여서 우선 RuntimeExcepion으로 처리 해놓은 상태입니다!
HttpService 연결 에러도 BudException으로 관리하는게 좋을까요?

Ggyumalang replied:
BudException이 RuntimeException을 상속받으니깐 BudException으로 해서 FAILED_GET_API_INFO? 뭐 이런식으로 해주면 좋을 것 같습니다!

Ggyumalang commented on user-api/src/main/java/zerobase/bud/news/service/NaverNewsApi.java:
```diff
@@ -0,0 +1,262 @@
+                ? mainImg.attr("data-src")
+                : "";
+
+        Element articleEle = getElementsExceptSizeZero(content, "#newsct_article")
```
"#newsct_article" 같은 상수들의 경우 private static final ARTICLE 뭐 이런 식으로 관리하거나 아니면 따로 클래스 News관련 클래스 만들어서 상수들을 관리하면 더 보기 좋고 관리하기가 편하더라구요! 한번 제안 드려봅니다.

xoals25 replied:
오오 좋은 꿀팁 감사합니다!
수정해서 올리도록 하겠습니다!

Ggyumalang reviewed (approved):
코드가 어렵고 길어보이는데 구현하시느라 너무 고생많으셨을 것 같습니다. 
코드가 긴대도 불구하고 잘 정리하신 것 같습니다!

thdefn commented on user-api/src/main/java/zerobase/bud/news/service/NaverNewsApi.java:
```diff
@@ -0,0 +1,262 @@
+    @Transactional
+    @Scheduled(cron = "0 0 0/1 * * *")
+    public void saveNaverNews() {
+        String[] keywords = {
```
이런 상수성 친구들은 Constant 클래스로 옮기면 참 좋을 것 같습니다 !

xoals25 replied:
꼼꼼하게 리뷰 봐주셔서 너무 감사합니다!
혹시 상수성 친구들은 keywords를 말씀하시는걸까요? 아니멸 cron = "0 0 0/1 * * *" 이부분을 말씀하시는걸까요?!

thdefn replied:
keywords 말한겁니다  ! 만들어둔 Constants 클래스가 있으니 거기로 옮겨주시면 좋을 것 같습니다

thdefn reviewed (approved):
저번에 로직이 많이 복잡해보였는데도 간결하게 잘 읽혀요 👍 정말 깔끔하게  구조를 잘 나누신 것 같습니다

xoals25 committed:
# Refactor : 코드 리뷰에서 받은 피드백대로 리펙토링

1. domain, repository 파일들 core로 이동
2. 접근 제한자 private 빼먹은 부분 추가
3. HttpService 에러 처리 BudException 으로 관리 할 수 있게 수정
4. 상수들은 Constant 클래스로 관리할 수 있게 수정
5. 기타 NaverNewsApi, NewsService 리펙토링

Sfer7 reviewed (approved):
고생하셨습니다!

xoals25 committed:
fix: 버그 수정

newsRepository 에서 테스트를 위해 수정한 부분을 다시 원래대로 돌렸습니다.

xoals25 committed:
test: 뉴스 테스트 코드 작성

NewsController 테스트 코드 작성
NewsService 테스트 코드 작성
adoc 문서 작성

xoals25 committed:
refactor: core로 이동한 domain, repository 다시 news 디렉토리로 이동

xoals25 committed:
feat: 스케줄러 시간 변경

xoals25 committed:
Merge branch 'develop' into feat#18
